### PR TITLE
fix: always set controlplane default envs

### LINKED ISSUE #63: Setting any controlplane env variable prevents default env vars from being created
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Currently default environment variables in controlplane object created set only if none of custom env variables were provided during controlplane object creation. Providing any custom env variables prevents defaults to from being created.

[controllers/controlplane_controller.go#L57](https://github.com/Kong/gateway-operator/blob/main/controllers/controlplane_controller.go#L57):

```go
debug(log, "validating ControlPlane configuration", controlplane)
if len(controlplane.Spec.Env) == 0 && len(controlplane.Spec.EnvFrom) == 0 {
	debug(log, "no ENV config found for ControlPlane resource, setting defaults", controlplane)
	setControlPlaneDefaults(&controlplane.Spec.ControlPlaneDeploymentOptions, controlplane.Namespace, nil)
	if err := r.Client.Update(ctx, controlplane); err != nil {
		if errors.IsConflict(err) {
			debug(log, "conflict found when updating ControlPlane resource, retrying", controlplane)
			return ctrl.Result{Requeue: true, RequeueAfter: requeueWithoutBackoff}, nil
		}
		return ctrl.Result{}, err
	}
	return ctrl.Result{}, nil // no need to requeue, the update will trigger.
}
```

It's not clear to me if this is intended or not.

### Expected Behavior

To be clarified.

#### Manual
```markdown
1. Set any env variable in spec of controlplane object.
2. See that defaults like `POD_NAME`, `POD_NAMESPACE`, `CONTROLLER_PUBLISH_SERVICE`, `CONTROLLER_KONG_ADMIN_URL` are not created.
```

#### Tests

1. Create a controlplane with any env var set
```go
controlplane := &operatorv1alpha1.ControlPlane{
	ObjectMeta: metav1.ObjectMeta{
		Namespace: controlplaneName.Namespace,
		Name:      controlplaneName.Name,
	},
	Spec: operatorv1alpha1.ControlPlaneSpec{
		ControlPlaneDeploymentOptions: operatorv1alpha1.ControlPlaneDeploymentOptions{
			DeploymentOptions: operatorv1alpha1.DeploymentOptions{
				Env: []corev1.EnvVar{{
					Name:  testEnvVar,
					Value: testEnvVal,
				}},
			},
			DataPlane: &nameOfDataPlane,
		},
	},
}
```

or if using gateway:
```go
gatewayConfig := &operatorv1alpha1.GatewayConfiguration{
	ObjectMeta: metav1.ObjectMeta{
		Namespace: namespace.Name,
		Name:      uuid.NewString(),
	},
	Spec: operatorv1alpha1.GatewayConfigurationSpec{
		DataPlaneDeploymentOptions: &operatorv1alpha1.DataPlaneDeploymentOptions{
			DeploymentOptions: operatorv1alpha1.DeploymentOptions{
				Env: []corev1.EnvVar{{
					Name:  testEnvVar,
					Value: testEnvVal,
				}},
			},
		},
		ControlPlaneDeploymentOptions: &operatorv1alpha1.ControlPlaneDeploymentOptions{
			DeploymentOptions: operatorv1alpha1.DeploymentOptions{
				Env: []corev1.EnvVar{{
					Name:  testEnvVar,
					Value: testEnvVal,
				}},
			},
		},
	},
}
```
2. Controlplane controller will hit the line [controllers/controlplane_controller.go#L57](https://github.com/Kong/gateway-operator/blob/main/controllers/controlplane_controller.go#L57)
3. Env variables are not set in control plane.
4. When a new deployment for control plane is created [controllers/controlplane_controller_utils.go#L69](https://github.com/Kong/gateway-operator/blob/4bf93e0c3390bdb71dba13fe0510c4f3d1607a5b/controllers/controlplane_controller_utils.go#L69), the empty ENV is copied [controllers/controlplane_controller_utils.go#L104](https://github.com/Kong/gateway-operator/blob/4bf93e0c3390bdb71dba13fe0510c4f3d1607a5b/controllers/controlplane_controller_utils.go#L104)
5. Pods created from the deployment do not have the `CONTROLLER_KONG_ADMIN_URL` set. As a result KIC will assume its default admin API endpoint.
6. Connection to admin API fail with log message `Error: could not retrieve Kong admin root: making HTTP request: Get "http://localhost:8001/": dial tcp [::1]:8001: connect: connection refused`
7. Other implications? `POD_NAME`, `POD_NAMESPACE`, `CONTROLLER_PUBLISH_SERVICE` are not set in pods too.

### Kong Ingress Controller version

```shell
N/A
```

### Kubernetes version

```shell
N/A
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by randmonkey**
**What this PR does / why we need it**:

always set default environment variables for `controlplane` resource, fix the bug that `POD_NAME` and `POD_NAMESPACE` is not set when any `env` specified in `ControlPlane`

**Which issue this PR fixes**

fixes #63

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [ ] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

File: controllers/controlplane_controller.go [hunk 1]
```diff
@@ -72,9 +72,12 @@ func (r *ControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	}
 
 	debug(log, "validating ControlPlane configuration", controlplane)
-	if len(controlplane.Spec.Env) == 0 && len(controlplane.Spec.EnvFrom) == 0 {
-		debug(log, "no ENV config found for ControlPlane resource, setting defaults", controlplane)
-		setControlPlaneDefaults(&controlplane.Spec.ControlPlaneDeploymentOptions, controlplane.Namespace, dataplaneServiceName, nil)
+	// TODO: add validating here
```
Comment (shaneutt) [inline controllers/controlplane_controller.go:75]: We need to resolve this TODO or have a follow up issue linked here.

**Review by shaneutt — CHANGES_REQUESTED** [review]
Looking good :+1: 

I think we're really close to being able to merge this, however given the context of https://github.com/Kong/gateway-operator/issues/63 it would appear now would be a good time to add something to our tests which proves we've resolved https://github.com/Kong/gateway-operator/issues/63 and helps prevent us from introducing regressions in the future.

I recommend we add (or even just update) an integration tests to validate the behavior of ENV var fulfillment when the user has or has not already provided vars.

File: controllers/controlplane_controller.go (hunk 1 above)
Comment (randmonkey) [reply controllers/controlplane_controller.go]: linked to #109.

File: controllers/controlplane_controller.go [hunk 2]
```diff
@@ -72,9 +72,12 @@ func (r *ControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Request
...
+
+	debug(log, "setting defaults for ControlPlane resource", controlplane)
```
[the remark this answers is not in the document]
Comment (randmonkey) [reply controllers/controlplane_controller.go]: done.

**Review by shaneutt — DISMISSED** [review]

**Comment by randmonkey** [comment]
Closed this and re-submit in #113 since signing for commits is required.

**Pull request closed without merging**